### ISSUE TITLE
Consider mirroring Google Fonts

### ISSUE THREAD

**Issue by notatestuser** [Synzvato/decentraleyes#34]
Plenty of sites use these. Could we mirror them in decentraleyes?

**Comment by Synzvato** [comment]
Thanks for the suggestion!

In theory yes. As of now, Decentraleyes does not support non-script files (e.g. fonts or styles). Once it does, it might indeed be an idea to start serving some of the web's most popular fonts.

Semi-related issues (see answers): #32, #37

**Comment by heforfree** [comment]
@Synzvato 
for font read this can be helpful

http://sosweetcreative.com/2613/font-face-and-base64-data-uri

**Comment by Synzvato** [comment]
@heforfree Thanks for sharing, will read!

**Comment by Gitoffthelawn** [comment]
Downloading from Google fonts is probably the most rampant.  At the same time, there are a few other common font CDN's that can be included as well.

**Comment by Gitoffthelawn** [comment]
@Synzvato Do you want to re-open this issue since #60 indicates that this may be a new feature?

Regardless, I have an interesting idea.  I've noticed that many times the font requested from a CDN may already be present on the system, yet is still downloaded by the browser.

Using javascript, it is trivial to detect some of the most commonly installed fonts.

As such, until decentraleyes explicitly supports fonts, it could simply force the browser to use it's own system's fonts instead of downloading practically duplicate copies (for fonts that are installed).  Yes, the system's fonts may not be absolutely 100% identical to the CDN fonts, but they are probably close enough for 99% of situations.

**Comment by Gitoffthelawn** [comment]
I've been testing the idea proposed in the above comment, and so far, it is working quite well.

**Comment by notatestuser** [comment]
Neat. Is there a clever way to deal with the fact that some custom font-faces use different names for system fonts? There can be subtle differences (missing spaces, added numbers, etc.) For example: HelveticaNeue

**Comment by Gitoffthelawn** [comment]
@notatestuser I'm not yet sure whether there is an integrated translation table like with PDF files.  From what I've tested so far, there is not.  However, writing one as an overlay is rather simple.

Fonts can also be edited with a hex editor to change their names.

Even with these tricks, it does not solve the subtle differences you mention.  From what I've tested so far, the impact of these subtle differences, fortunately, is negligible.  I haven't found any issues yet, but I've only tested a few hundred sites... and the web has a few more sites than that.  ;-)

**Comment by AshotN** [comment]
Any updates in regards to this?

**Comment by Synzvato** [comment]
@AshotN Resource additions are temporarily blocked by ([```1419459```](https://bugzilla.mozilla.org/show_bug.cgi?id=1419459)). Please see #16 for more details.

**Comment by henri98** [comment]
Any updates on this?

**Comment by CHEF-KOCH** [comment]
I'm against this implementation [Google's fonts are already open source](https://github.com/google/fonts) and integrating it as an own resource only causes more resource been wasted for no benefit. In the meantime they could already have changed the fonts so you would be forced to keep an eye on it but even if you allow them externaly I see no reason how it compromises your privacy, it's well explained [here](https://developers.google.com/fonts/faq#Privacy). I think if you read it and compare this against GitHub and other pages it doesn't collect more information, I also haven't found any evidence that tracking fonts have any impact on security - it's external CDN which supports the tracking code, a font by itself doesn't track anything. The _recently_ holes which got discovered back in 2015 are [compromised by a remote code execution](https://docs.microsoft.com/en-us/security-updates/securitybulletins/2015/ms15-078). However. this techniques is in general a security problem and not only relevant to fonts. 

If you don't like fonts you can already block them permanently or temporarily within uBlock/AdGuard etc.  If you just want to obfuscate from which domain you're coming you can set like several years now the [noreferrer tag](http://www.w3.org/TR/html5/links.html#rel-noreferrer) which also works in HTML5. 

The possible tracking which people are often refer too are not the fonts, it is the fact that the page you like to visit or coming from like to [set the cookies](https://en.wikipedia.org/wiki/HTTP_ETag). So I don't see here what are the privacy benefits. I personally don't know what is wrong with the fonts which the Browser already includes, should't that be prefered compared to any external (CDN/addon integrated ones)?

**Comment by ronjouch** [comment]
@CHEF-KOCH 

> *"Google's fonts are already open source"*
> [...]
> *"they could already have changed the fonts so you would be forced to keep an eye on it"*

That's unrelated to the present discussion.

> *"integrating it as an own resource only causes more resource been wasted"*

If you take this stance, then all Decentraleyes should go to the wastebasket. Decentraleyes *already* takes the trade-off of an initial cost of pre-packaged commonly-used web resources vs. fetching them on demand.

> *"even if you allow them externaly I see no reason how it compromises your privacy, it's well explained [here](https://developers.google.com/fonts/faq#what_does_using_the_google_fonts_api_mean_for_the_privacy_of_my_users)"*

That's purely Google's word, can change at any time, can be enforced by governement mandates, and many netizens simply don't trust it.

> *"If you don't like fonts you can already block them permanently or temporarily within uBlock/AdGuard etc. "*

Strawman (you're criticizing something something different from what is being argued). I (and users interested in this feature) certainly don't want to block fonts. Of course we could if we wanted to. This feature is precisely about *preserving* webfont UX. I *do* like neat fonts and take the crisp serifs served by Gfonts rather than stock OS fonts. Also, some sites may *depend* on glyphs present in these fonts and not in my OS fonts, meaning the loss due to blocking would be more than just visual.

> *"The possible tracking which people are often refer too are not the fonts, it is the fact that the page you like to visit or coming from like to set the cookies"*

There are many more ways to track people on the internet (see [EFF's panopticlick](https://panopticlick.eff.org/)). The threat with fonts may be mild, but at the most fundamental level, I don't especially like the idea of me requesting resources (logged by Google servers/CDNs) on half the web.

---

And I'm done contradicting, now to a simple & short positive assertion: packaging GoogleFonts in Decentraleyes would be valuable for the same reasons as packaging js libs is: performance and privacy.

I hope it can land someday 🙂.

**Comment by CHEF-KOCH** [comment]
> That's unrelated to the present discussion.

It's relevant since the web changes frequently. Font rendering got improved already since the OP started this discussion here. [Different fonts](http://brick.im/preview/brick.html) [look different in the same browser](http://brick.im/preview/google.html). 

> If you take this stance, then all Decentraleyes should go to the wastebasket. Decentraleyes already takes the trade-off of an initial cost of pre-packaged commonly-used web resources vs. fetching them on demand. Also, note the adjective contained in this issue's title: "Add popular Google Fonts to the bundle", nobody's asking for a 300MB xpi 🙂, same as Decentraleyes already packages a chosen subset of js libs..

The OP nor you mention 'which' fonts exactly, if you do as the OP says it would exactly ends-up with a resource wasting addon without benefit. I see no argument to globally disallow fonts in e.g. Firefox (again what's wrong with the Browser ones?) - or to block tem with known extensions like uBo.

> That's purely Google's word, can change at any time, can be enforced by governement mandates, and many netizens simply don't trust it.

Bring evidence, there is none that fonts by default are a security risk, it's not possible. The mentioned example from me points to a weak font driver which got exposed by remote code execution, that also affects Browser by itself and isn't a proof at all that the fonts are somewhat a security issue at all. The possible request to it might could be MITM (in theory) but to stay on the topic, Google's ones are delivered via HTTPS by default (since years). 

> Strawman (you're criticizing something something different from what is being argued). I (and users interested in this feature) certainly don't want to block fonts. Of course we could if we wanted to. This feature is precisely about preserving webfont UX. I do like neat fonts and take the crisp serifs served by Gfonts rather than stock OS fonts. Also, some sites may depend on glyphs present in these fonts and not in my OS fonts, meaning the loss due to blocking would be more than just visual.

I not even talked to you but okay. I was refering to OP which is should be the matter here. If you don't want to block fonts why replace them with others which are locally used? Doesn't make much sense when the Browser (as said) already includes all nessary fonts by default. There is then same like you would block it no network bandwith 'wasted' and the same benefit are given except that you need an extension/addon to do that. 

It's nice that you tell me which fonts you prefer but doesn't have to do anything with 'Google fonts'. 

> There are many more ways to track people on the internet (see EFF's panopticlick). The threat with fonts may be mild, but at the most fundamental level, I don't especially like the idea of me requesting resources (logged by Google servers/CDNs) on half the web.

There is no font threat at all and there never was as I already explained. EFF also not give any proof how font detection has an security impact, all they do is to tell people it might expose metadata, yes we know it already but no one came up woth POC to show that this exposes you. If I post my Browser fonts in GitHub in public without context you think I'm insecure? Wrong, posting a comment here exposes me more - as a matter of fact.

> And I'm done contradicting, now to a simple & short positive assertion: packaging GoogleFonts in Decentraleyes would be valuable for the same reasons as packaging js libs is: performance and privacy.

Packing tons of fonts into an extension doesn't result in better performance not any privacy improvemt. Please bring evidence. And I bring the counter evidence that my debug console shows more memory consumption (which can also expose you according to the same EFF logic) and an article that CDN load the tracking scripts not the fonts by itself.

**Comment by crssi** [comment]
@CHEF-KOCH
Don't you think that people does might not see fonts as a security problem, but as a tracking problem.
I personally like Decentraleyes for de-tracking reasons.
What is really your intentions here exactly?

**Comment by Synzvato** [comment]
@henri98, @ronjouch

> I hope it can land someday [...].

I'm happy to say that addressing the underlying technical issues is top priority. Once that is done, adding support for popular webfonts should not take too long. In short, this feature is still on the roadmap.

**Comment by Gitoffthelawn** [comment]
BTW, according to [Google itself](https://developers.google.com/fonts/faq#what_does_using_the_google_fonts_api_mean_for_the_privacy_of_my_users), ***"Google Fonts logs records of the CSS and the font file requests..."*** each and every time a person requests an uncached font from a Google server.

**Comment by shellshocker** [comment]
@CHEF-KOCH
@All

Just ignore this CHEF-KOCH. Mostly he does not know what he is talking about. He just produces text walls in different projects.

**Comment by CHEF-KOCH** [comment]
> BTW, according to Google itself, "Google Fonts logs records of the CSS and the font file requests..." each and every time a person requests an uncached font from a Google server.

That's copy & paste from what I already linked, I already explains that the font's itself doesn't do anything. The log processing here comes (in this case) from google. Once Google Javascript is blocked or the domain locally loaded nothing will be logged.

From this comment [here](https://github.com/Synzvato/decentraleyes/issues/274).

> Just note that Decentraleyes focuses on privacy rather than saving bandwidth, so solving this issue will most likely have lower priority.

Indeed and the ones who have no clue are people without proper research. Fonts itself do nothing and there open source.

**Comment by shellshocker** [comment]
Once again: Just ignore this guy (@CHEF-KOCH). He does not know what he is talking about.

**Comment by CHEF-KOCH** [comment]
From the research paper [Fingerprinting web users through font metrics](https://github.com/CHEF-KOCH/Online-Privacy-Test-Resource-List/raw/master/Documents/Font/Fingerprinting%20web%20users%20through%20font%20metrics.pdf). Keep in mind that this document is already _outdated_.   It should also be mentioned that this requies a domain/site which uses actievly an exploit/attack related to identify you. It's like using Panopticlick to show hey I'm possible vulnerable ... It's same logic like driving against a tree to proof that this is possible dangerous. In my eyes that should't be the goal at all. 

```
The attack as we have designed it relies on client-side execution of JavaScript.
Simply disabling JavaScript is unlikely to be an effective defense, however. Hei-
derich et al. [11] show how to use CSS to measure the size of DOM elements, for
example by shrinking a container through animation and causing its contents to
reflow.
```

```
A simple idea to eliminate variation due to font file availability is to ship a
set of standard fonts with the web browser, and use only those (plus download-
able web fonts), at the exclusion of any other fonts that may be installed on
the system. This approach has been suggested by Mowery and Schacham [16,
Section 5] and on the bug trackers of Mozilla [22] and Tor [21]
```

Once again disabling JS/Cookies etc is already enough. I already said that there is nothing wrong with using Browser integrated fonts or using uBO, both have same bits 8 (you need 8+ bits in order to reveal someones identify [according to EFF wrong paper] [[cause their mathematical explanation doesn't respect several aspects like false positive while spoofing]] so in reality you need 10 bits). Even if I enable it and allow JS I currently see max 9,2 bits which is noth enough to expose me.

It's same story like with favicons, they don't expose anything while downloading, it's the curcumstances like placing the cookies or trying to fingerprint the connection itselt. Fighting metadata is simply a mission impossible.

**Comment by notatestuser** [comment]
Thanks for participating in my thread, but this is simple - your IP and browser fingerprint are exposed because the font file must be downloaded from their servers. We’re not talking about JS execution. That’s the end of the discussion about this.

**Comment by shellshocker** [comment]
I told you. He (@CHEF-KOCH) does not know what he is talking about. It's simple: Start ignoring his bullshit. In other projects he is doing the same. It's really bad. -.-
